src/Flysystem/FlysystemAssetStore.php

```diff
@@ -785,9 +798,37 @@ protected function parseFileID($fileID)
 
         $filename = $matches['folder'] . $matches['basename'] . $matches['extension'];
         $variant = isset($matches['variant']) ? $matches['variant'] : null;
+        $hash = isset($matches['hash']) ? $matches['hash'] : null;
         return [
             'Filename' => $filename,
             'Variant' => $variant,
+            'Hash' => $hash
+        ];
+    }
+
+    /**
+     * Try to parse a file ID using the old SilverStripe 3 format legacy or the SS4 legacy filename format.
+     *
+     * @param string $fileID
+     * @return array
+     */
+    private function parseLegacyFileID($fileID)
+    {
+        // assets/folder/_resampled/ResizedImageWzEwMCwxMzNd/basename.extension
+        $ss3Pattern = '#^(?<folder>([^/_]+/)*?)(_resampled/(?<variant>([^/.]+))/)?((?<basename>((?<!__)[^/.])+))(?<extension>(\..+)*)$#';
```

Comment:
How much did you prod and poke at those regexes for edge cases?

Reply:
Added a bunch of extra unit tests.